src/cdk/scrolling/for-of.ts

```diff
@@ -0,0 +1,275 @@
+      for (let i = 0, ilen = view.rootNodes.length; i < ilen; i++) {
+        let rect = (view.rootNodes[i] as Element).getBoundingClientRect();
+        minTop = Math.min(minTop, rect.top);
+        minLeft = Math.min(minLeft, rect.left);
+        maxBottom = Math.max(maxBottom, rect.bottom);
+        maxRight = Math.max(maxRight, rect.right);
+      }
+
+      return {
+        top: minTop,
+        left: minLeft,
+        bottom: maxBottom,
+        right: maxRight,
+        height: maxBottom - minTop,
+        width: maxRight - minLeft
+      };
+    }
+    return null;
+  }
+
+  ngDoCheck() {
+    if (this._differ && this._needsUpdate) {
+      const changes = this._differ.diff(this._renderedItems);
+      this._applyChanges(changes);
+      this._needsUpdate = false;
+    }
+  }
+
+  ngOnDestroy() {
+    this._viewport.disconnect();
+
+    this._dataSourceSubject.complete();
+    this.viewChange.complete();
+
+    for (let view of this._templateCache) {
+      view.destroy();
+    }
+  }
+
+  /** React to scroll state changes in the viewport. */
+  private _onRenderedRangeChange(renderedRange: Range) {
+    this._renderedRange = renderedRange;
+    this.viewChange.next(this._renderedRange);
+    this._renderedItems = this._data.slice(this._renderedRange.start, this._renderedRange.end);
+    if (!this._differ) {
+      this._differ = this._differs.find(this._renderedItems).create(this.cdkForTrackBy);
+    }
+    this._needsUpdate = true;
+  }
+
+  /** Swap out one `DataSource` for another. */
+  private _changeDataSource(oldDs: DataSource<T> | null, newDs: DataSource<T>): Observable<T[]> {
+    if (oldDs) {
+      oldDs.disconnect(this);
+    }
+    this._needsUpdate = true;
+    return newDs.connect(this);
+  }
+
+  /** Apply changes to the DOM. */
+  private _applyChanges(changes: IterableChanges<T> | null) {
+    // If there are no changes, just update the index and count on the view context and be done.
+    if (!changes) {
+      for (let i = 0, len = this._viewContainerRef.length; i < len; i++) {
+        let view = this._viewContainerRef.get(i) as EmbeddedViewRef<CdkForOfContext<T>>;
+        view.context.index = this._renderedRange.start + i;
+        view.context.count = this._data.length;
+        view.detectChanges();
+      }
+      return;
+    }
+
+    // Detach all of the views and add them into an array to preserve their original order.
+    const previousViews: EmbeddedViewRef<CdkForOfContext<T>>[] = [];
+    for (let i = 0, len = this._viewContainerRef.length; i < len; i++) {
+      previousViews.unshift(
+          this._viewContainerRef.detach()! as EmbeddedViewRef<CdkForOfContext<T>>);
+    }
+
+    // Mark the removed indices so we can recycle their views.
+    changes.forEachRemovedItem(record => {
+      this._templateCache.push(previousViews[record.previousIndex!]);
+      delete previousViews[record.previousIndex!];
```

Comment:
Using `delete` here won't reindex the array, is that intended? Otherwise it might be better to use `splice` to avoid some unintended side-effects. Same goes for the other `delete` usages.

Reply:
Yes this is intentional. I need each element to be at its correct original index so it matches up with the index in the IterableChangeRecord. I then use `delete` to mark the view as having been handled in some way (cached or queued up for re-insertion).

Reply:
Would it work to to `null`? That's generally much faster than using `delete`